### PR TITLE
tinyformat: enforce compile-time checks for format string literals

### PR DESCRIPTION
**Pull request by stickies-v**
_Carved out of #30928 - this PR doesn't include the controversial change of suppressing `tinyformat::format_error` throwing behaviour, and just adds enforcement of compile-time checks for format string literals._

Reduce unexpected run-time crashes from string formatting by enforcing compile-time format string checks for most* `tfm::format` and `strprintf` usage.

This PR should introduce no behaviour change.
The main changes are:
- remove the `const std::string&` `tfm::format` overload since it's not necessary. Usage of this overload is removed by one of:
  - replacing string concatenation in the format string with just an extra parameter
  - using the `bilingual_str` overload
  - using the new `tfm::format_raw` functions (only for tinyformat implementation or tests)
- rename the non-compile-time-validated `tfm::format` overloads to `tfm::format_raw`, so existing callsites by default use the safer `util::ConstevalFormatString` `tfm::format` overloads. Callsites that for some reason don't pass the compile-time checks (such as in `bitcoin-cli.cpp`) can use `tfm::format_raw`.

See https://github.com/bitcoin/bitcoin/pull/31149#discussion_r1816650254 for an example on how this PR can prevent a run-time error for a small format string mistake.

_* See #31061 for `bilingual_str format(const bilingual_str& fmt, const Args&... args)` compile-time checks._

### CODE REVIEW

**Review by hodlinator — APPROVED** [review]
ACK 4cf12216951e91550c81db807fe0ecfafe6834c9

Only dropped commit I had [concerns about](https://github.com/bitcoin/bitcoin/pull/30928#issuecomment-2431257072). [l0rinc too](https://github.com/bitcoin/bitcoin/pull/30928#pullrequestreview-2389270316).

Does not entirely prevent throwing of exceptions from incorrect format strings (notably translated strings), but is a step in the right direction (validating more format strings at compile time).

File: src/test/fuzz/strprintf.cpp
Comment (stickies-v) [reply src/test/fuzz/strprintf.cpp]: Good idea, thanks - added you as co-author to 99b12b5cfcf438a59e9f54ad6d2228be86a9bfeb

**Review by stickies-v — COMMENTED** [review]
Rebased to address merge-conflict from #29936. Also added commit 99b12b5cfcf438a59e9f54ad6d2228be86a9bfeb to minimize `c_str()` usage (an [earlier review concern](https://github.com/bitcoin/bitcoin/pull/30928#issuecomment-2389355374) of @ryanofsky) and minimize code churn, as [suggested](https://github.com/bitcoin/bitcoin/pull/31149#discussion_r1816298165) by @maflcko.

File: src/bitcoin-cli.cpp
Comment (jonatack) [inline src/bitcoin-cli.cpp:546]: Is all the indentation churn really necessary?

Does changing strprintf to tfm_format_raw here catch any potential issues that wouldn't be seen in manual testing?

Comment (stickies-v) [reply src/bitcoin-cli.cpp:546]: > Is all the indentation churn really necessary?

It's `clang-format` output which I believe is what we're trying to adhere to.

> Does changing strprintf to tfm_format_raw here catch any potential issues that wouldn't be seen in manual testing?

The code does not compile with `strprintf` because this PR enforces compile-time checks, which fail because of the width specifiers in the format string (there's quite some discussion on these limitations in https://github.com/bitcoin/bitcoin/pull/30546). See also this excerpt from the PR description that addresses this:

> * rename the non-compile-time-validated `tfm::format` overloads to `tfm::format_raw`, so existing callsites by default use the safer `util::ConstevalFormatString` `tfm::format` overloads. Callsites that for some reason don't pass the compile-time checks (such as in `bitcoin-cli.cpp`) can use `tfm::format_raw`.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: Clang-format can be helpful for new code, but I don't think it's necessary to re-indent code for it that isn't otherwise touched -- that's just noisy churn.

Comment (stickies-v) [reply src/bitcoin-cli.cpp:546]: > Clang-format can be helpful for new code, but I don't think it's necessary to re-indent code for it that isn't otherwise touched -- that's just noisy churn.

I'm just going with our `clang-format` settings as defined in `src/.clang-format` - if you feel like those settings are suboptimal opening a separate pull/issue for that is probably more appropriate?

> Please provide an example of how this code could crash currently without this change?

The PR states that there is no behaviour change, so (unless there are run-time bugs we haven't seen yet) we can't make this code crash without this change. The point of this PR is to make the code more robust against future changes, where a new/modified format string can cause a run-time error, e.g. if not all code-paths are manually tested by reviewers (which is not unusual).

Comment (stickies-v) [reply src/bitcoin-cli.cpp:546]: > Please provide an example of how this code could crash currently without this change?

If you're asking about how such an error could sneak in without this PR, a change as simple adding a single extra `*` would lead to a run-time exception if no extra parameter is added:

```cpp
            result += tfm::format_raw("<->   type   net  v  mping   ping send recv  txn  blk  hb %**s%*s%*s ",
```

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: Thanks for the example. As the primary author of this code, I think the run-time error would be more clear, though a compile-time error would catch it one CLI command earlier. I haven't found this kind of error to be an issue for me over the past half decade on this code.

Also, this code path is manually tested by myself, reviewers and users. Perhaps for that reason, and because it's non-critical code, a maintainer explicitly asked that no test coverage be written for it.

So either way is fine, I suppose, but this conflicts with https://github.com/bitcoin/bitcoin/pull/30930 so it's more an annoyance than helpful for me.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: > I'm just going with our clang-format settings as defined in src/.clang-format

It's not a hard requirement to retouch existing code with that tool, and I'd prefer avoiding the churn.

Comment (stickies-v) [reply src/bitcoin-cli.cpp:546]: > As the primary author of this code, I think the run-time error would be more clear

I'm sorry, what? For what reason would we not use the logic we already have to catch (some/most) of these format errors at compile-time, and defer them to run-time instead, relying on manual review to make sure each code-path is ran, potentially crashing user software? That seems like the opposite of security engineering best practices to me.

> a maintainer explicitly asked that no test coverage be written for it.

We're not writing any test coverage for it here either, it comes for free. How is that a bad thing?

> It's not a hard requirement to retouch existing code with that tool, and I'd prefer avoiding the churn.

Noted, but I prefer keeping this as is, I'd rather not override `clang-format` for this.

**Review by sedited — APPROVED** [review]
Nice, ACK 23560b468c474bbc6a3abd6c0778da62766ac8f7

File: src/bitcoin-cli.cpp
Comment (TheCharlatan) [reply src/bitcoin-cli.cpp:546]: > I'd prefer avoiding the churn.

We usually call things churn that change things multiple times within the same PR. Indenting the lines below a changed line that belong to the same logical unit is a good thing in my eyes. If we would not do that anywhere, the code would look weird and probably make `clang-format` unusable after a while.

Comment (stickies-v) [reply src/bitcoin-cli.cpp:546]: > The rest of this PR seems fine.

We can't have the rest of this PR without updating `bitcoin-cli.cpp`, unless someone implements the complete format string validation logic, for which as per my earlier comment, see https://github.com/bitcoin/bitcoin/pull/30546.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: > I'm sorry, what? For what reason would we not use the logic we already have to catch (some/most) of these format errors at compile-time, and defer them to run-time instead, relying on manual review to make sure each code-path is ran, potentially crashing user software? 

> relying on manual review to make sure each code-path is ran, potentially crashing user software

Because, as I wrote:

- I think the run-time error is more clear to understand, as the person working on the code
- This code path is manually tested anyway, otherwise the command with any non-zero level won't run

> That seems like the opposite of security engineering best practices to me.

One could argue the same for discouraging test coverage, but that's been the stance.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: > > I'd prefer avoiding the churn.
> 
> We usually call things churn that change things multiple times within the same PR. Indenting the lines below a changed line that belong to the same logical unit is a good thing in my eyes. If we would not do that anywhere, the code would look weird and probably make `clang-format` unusable after a while.

You're changing those lines needlessly. Whether you prefer to call it noise or churn is unimportant.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: @TheCharlatan mind being explicit what part of my feedback your emoji relates to?

Comment (stickies-v) [reply src/bitcoin-cli.cpp:546]: > I think the run-time error is more clear to understand, as the person working on the code

And I think prioritizing devs getting nicer error messages over increased software run-time stability is almost never a good trade-off. It seems we have fundamentally different views, so I'll leave this thread unresolved but I strongly disagree with your approach and won't change the PR because of it.

> This code path is manually tested anyway, otherwise the command with any non-zero level won't run

I understand it's tested, and that's good, because `tfm::format_raw()` doesn't do any compile-time checks.

Comment (TheCharlatan) [reply src/bitcoin-cli.cpp:546]: > mind being explicit what part of my feedback your emoji relates to?

Sure

> I haven't found this kind of error to be an issue for me over the past half decade on this code.

I've seen a bunch of positional args errors in fairly recent times, and catching them at compile time is certainly worthwhile. People tried doing this with various linters, but the linters produced false-positives, or missed certain cases.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: > And I think prioritizing devs getting nicer error messages over increased software run-time stability is almost never a good trade-off. It seems we have fundamentally different views

I think we agree there, given that you wrote "almost" ;)

**Review by hodlinator — APPROVED** [review]
re-ACK 23560b468c474bbc6a3abd6c0778da62766ac8f7

Changes since [previous ACK](https://github.com/bitcoin/bitcoin/pull/31149#pullrequestreview-2393618400):
- Added file-local `fuzz_fmt` function to reduce churn.
- Rebased and applied conversions to new code from #29936.

File: src/bitcoin-cli.cpp
Comment (l0rinc) [reply src/bitcoin-cli.cpp:546]: > Does changing strprintf to tfm_format_raw here catch any potential issues that wouldn't be seen in manual testing?

By manual testing do you mean automated testing of print statements?

> Clang-format can be helpful for new code, but I don't think it's necessary to re-indent code for it that isn't otherwise touched

Please format affected code, don't leave it unformatted, that's why the current code is so confusing.
```
Coding Style (General)
----------------------

Various coding styles have been used during the history of the codebase,
and the result is not very consistent. However, we're now trying to converge to
a single style, which is specified below. When writing patches, favor the new
style over attempting to mimic the surrounding style, except for move-only
commits.
```

> I haven't found this kind of error to be an issue for me over the past half decade on this code.

Others have found it to be an issue, better catch them as early as we can. Why wait until runtime or CI failure or release.

Comment (l0rinc) [reply src/bitcoin-cli.cpp:546]: There was no bikeshedding about this before on https://github.com/bitcoin/bitcoin/pull/31149, it just started.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: > By manual testing do you mean automated testing of print statements?

No, manual testing by building and running the CLI commands.

There is no automated test coverage of `-netinfo` (the touched code above), nor, for instance, of `-addrinfo`.

**Review by l0rinc — APPROVED** [review]
Compared to https://github.com/bitcoin/bitcoin/pull/30928 I see the following changes:
* **clientversion.cpp:** Merge conflict in clientversion at CopyrightHolders -> seems to be resolved correctly
* **logging.h:** old LogPrintFormatInternal is retained
* **run-lint-format-strings.py:** FALSE_POSITIVES not removed
* **util_string_tests.cpp:** FALSE_POSITIVES not moved to tests
* **strprintf.cpp:** fuzz_fmt added to avoid renames, try-catch added around fuzz_fmt calls (any reason for not try-catching inside fuzz_fmt instead?)
* **tinyformat.h:** try-catch removed from vformat
* **wallet.h:** new strprintf migrated

ACK b448d6082674e2c6d2a172145e33795b01e072ff

File: src/bitcoin-cli.cpp
Comment (l0rinc) [reply src/bitcoin-cli.cpp:546]: By doing some of the validation at compile time, there **is** now some automated 'test coverage' on it - i.e. it will bite back if it's very wrong without any additional manual check.

Comment (jonatack) [reply src/bitcoin-cli.cpp:546]: Generally unneeded changes can be avoided, but that was just a suggestion. Further historical context to get started:

- https://github.com/bitcoin/bitcoin/issues/18651
- https://github.com/bitcoin/bitcoin/issues/15465

Comment (l0rinc) [reply src/bitcoin-cli.cpp:546]: Useful context, thanks!
We must however keep our kitchen clean, we can't just be cooking all the time.
This change by @stickies-v was a small, low-risk change, making the code slightly more maintainable (or at least slightly less unmaintainable).
We should encourage these.

**Comment by ryanofsky** [comment]
Sorry for causing unnecessary churn with my earlier comments. After reading Marco's comment in https://github.com/bitcoin/bitcoin/pull/31072#discussion_r1801085975 I finally understand what this PR is really trying to do, and why all the changes are being made here. Before reading that comment I thought this PR was trying not just to add compile-time checking for tinyformat format strings, but also to force calls that don't use compile-time checking to use different formatting functions like `format_raw` and `fuzz_fmt`. But now I see goal of this PR really is to enable compile-time checking, and the other changes here are basically just a side-effect of the approach taken to implement it.

That being said, I think the approach in this PR is not ideal. I think introduction of `_raw()` functions and uses of `c_str()` are ugly and unnecessary, and even though there are good changes here like code cleanup in various places, those changes should not be necessary just to have compile-time checking of literal format strings, and bundled in with less positive changes here.

I implemented an different approach in #31174 which only adds compile-time checking for `strprintf`/`tfm::format` string literals, and doesn't make other changes. If that PR looks ok, maybe we could use that approach, and then sort out which other changes here are actually improvements on their own terms? I would much prefer that to bundling good, bad, and ugly changes together in one PR.

**Comment by ryanofsky** [comment]
> I think there is merit in making the less safe (i.e. `std::string` overload) less convenient to use so that developers who are less familiar with this code don't accidentally use it when they don't have to, but I'll leave a comment on your PR so we can have the conversation there.

I agree with this, and commit b6a39c81e85338bc82f3db924157a599aa7e25fa in this PR shows places where code is doing things like `strprintf("[" + comment + "]")` that get around compile time checking. It'd be great to keep the changes in this PR which clean up that code, and add a more explicit syntax for passing unchecked strings. I think I would prefer a syntax more like `strprintf(Unchecked(desc_fmt), ...)` instead of `tfm::format_raw(desc_fmt.c_str(), ...)`, but I mainly just think it is best to make these changes separately so PRs are easier to understand and don't touch so much code.

**Pull request closed without merging**
